### PR TITLE
Add compilers paths for dependents of intel-parallel-studio

### LINKED ISSUE #2978: scotch does not depend on intelmpi
Hi,

I'm trying to install an intelmpi version of scotch (just installed an openmpi version) with
`spack spec scotch+esmumps+mpi ^intelmpi` which fails with `==> Error: scotch does not depend on intelmpi
`, while the openmpi version gives
```bash
$ spack spec scotch+esmumps+mpi ^openmpi
Input spec
--------------------------------
scotch+esmumps+mpi
    ^openmpi

Normalized
--------------------------------
scotch+esmumps+mpi
    ^bison
        ^m4@1.4.6:
    ^flex@:2.6.1
        ^gettext@0.19:
        ^help2man
    ^openmpi
        ^hwloc
            ^libpciaccess
                ^libtool
                ^pkg-config@0.9.0:
                ^util-macros

Concretized
--------------------------------
scotch@6.0.4%gcc@6.1.0+compression+esmumps+metis+mpi+shared arch=linux-centos7-x86_64 
    ^bison@2.7%gcc@6.1.0 arch=linux-centos7-x86_64 
    ^flex@2.5.37%gcc@6.1.0 arch=linux-centos7-x86_64 
    ^openmpi@1.10.3%gcc@6.1.0~java~mxm~pmi~psm~psm2~slurm~sqlite3~thread_multiple~tm~verbs+vt arch=linux-centos7-x86_64 
    ^zlib@1.2.8%gcc@6.1.0+pic+shared arch=linux-centos7-x86_64 

```

I do not understand why I cannot build with intelmpi (and intel compilers) here.
Relevant part of packages.yaml:
```yaml
packages:
  openmpi:
    modules:
      openmpi@1.10.3%gcc@6.1.0: openmpi/1-10.3--gnu--6.1.0
    buildable: False
  intelmpi:
    modules:
      intelmpi@2017.1.132%intel@17.0.1: intelmpi/2017--binary
    buildable: False
  zlib:
    modules:
      zlib@1.2.8%gcc@6.1.0: zlib/1.2.8--gnu--6.1.0
    buildable: False
  all:
    providers:
      mpi: [openmpi, intelmpi]
```

I have one guess: is this happening because zlib is only available in a gcc version?
Then I would find this to be a very cryptic error message.

### PR DESCRIPTION
**Pull request by Exteris**
Fixes #2978

### CODE REVIEW

**Comment by Zzzoom** [comment]
Is this just for Intel MPI or the whole Intel toolchain?
mpi{icc,icpc,ifort} will always run intel compilers, but if you want to use Intel MPI with any other compiler you should set I_MPI_CC and run mpicc.

**Comment by davydden** [comment]
> mpi{icc,icpc,ifort} will always run intel compilers, but if you want to use Intel MPI with any other compiler you should set I_MPI_CC and run mpicc.

that's a valid point. We should also make sure Spack's compiler wrappers are used underneath. For `openmpi` this is done via
```
        spack_env.set('OMPI_CC', spack_cc)
        spack_env.set('OMPI_CXX', spack_cxx)
        spack_env.set('OMPI_FC', spack_fc)
        spack_env.set('OMPI_F77', spack_f77)
```
in `setup_dependent_environment`. 

@Zzzoom i never used Intel MPI, so does it always have `mpicc` in bins?

**Comment by Zzzoom** [comment]
I can see both mpicc and mpiicc in my IPSXE 2016u4 mpi/bin64 directory.

**Comment by davydden** [comment]
@lee218llnl if I am not mistaken, you contributed `intel` packages, do you know about differences between `mpicc` and `mpiicc` and how we should add these wrappers in Spack?

**Comment by davydden** [comment]
@Exteris @Zzzoom it looks like we can use `mpicc` with intel compilers as well:

> Note that if you want to use mpif90/mpicc/etc (eg for existing configure scripts/Makefiles, etc) with the Intel compilers, one need only modify environment variables to point to the appropriate compilers: eg,
export I_MPI_CC=icc
export I_MPI_CXX=icpc
export I_MPI_F77=ifort
export I_MPI_F90=ifort

from https://software.intel.com/en-us/forums/intel-clusters-and-hpc-technology/topic/288354

So maybe one just need to add
```
spack_env.set('I_MPI_CC', spack_cc)
spack_env.set('I_MPI_CXX', spack_cxx)
spack_env.set('I_MPI_F77', spack_fc)
spack_env.set('I_MPI_F90', spack_f77)
```
together with standard wrapper names (i.e. `mpicc`).

**Comment by lee218llnl** [comment]
@davydden it looks like you answered your own question. By default the mpicc, mpicxx, mpif, etc. scripts will use the GNU compilers unless you set the I_MPI_* variables. Note you should also add:

spack_env.set('I_MPI_FC', spack_fc)

**Comment by davydden** [comment]
@lee218llnl thanks for confirming.

@Exteris would you mind to update the PR accordingly?

**Comment by Exteris** [comment]
Updated.
I need to stress for this PR that I have not tested this with a spack-installed intel-parallel-studio, but only with intel-parallel-studio loaded from a module. Perhaps someone who installs their own IPSXE can try?

File: var/spack/repos/builtin/packages/intel-parallel-studio/package.py [hunk 1]
```diff
@@ -305,6 +305,13 @@ def setup_environment(self, spack_env, run_env):
             # MPI. It is also possible to set the variables in the modules.yaml
             # file if Intel MPI is the dominant, or only, MPI on a system.
             run_env.set('I_MPI_ROOT', join_path(self.prefix, 'impi'))
+            # Set mpi environment variables
+            spack_env.set('I_MPI_CC', spack_cc)
```
Comment (davydden) [inline var/spack/repos/builtin/packages/intel-parallel-studio/package.py]: I am not sure but in `openmpi` these things are done in `def setup_dependent_environment(self, spack_env, run_env, dependent_spec):`.  see the docs http://spack.readthedocs.io/en/latest/packaging_guide.html?highlight=setup_dependent_environment
`setup_environment` appears to be mostly modules related http://spack.readthedocs.io/en/latest/module_file_support.html?highlight=setup_environment

Comment (Exteris) [inline var/spack/repos/builtin/packages/intel-parallel-studio/package.py]: You are probably right, I'll alter it.
What is the function signature? In the docs I see both:
```python
    def setup_dependent_environment(self, module, spec, dep_spec):
        """Dependencies of Qt find it using the QTDIR environment variable."""
        os.environ['QTDIR'] = self.prefix
```

The arguments to this function are:

    module: the module of the dependent package, where global properties can be assigned.
    spec: the spec of the dependency package (the one the function is called on).
    dep_spec: the spec of the dependent package (i.e. dep_spec depends on spec).

and
```python
def setup_dependent_environment(self, spack_env, run_env, extension_spec):
```

The difference in names confuses me.

Comment (davydden) [inline var/spack/repos/builtin/packages/intel-parallel-studio/package.py]: i would take the same one which is used in `openmpi`:
```
def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
```

**Comment by davydden** [comment]
@Exteris what's the status of this PR? would you mind fixing the minor issue related to `setup_dependent_environment`?

File: var/spack/repos/builtin/packages/intel-parallel-studio/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/intel-parallel-studio/package.py]: I fixed the inconsistencies in the documentation in #3351. You should use:
```python
setup_dependent_environment(self, spack_env, run_env, dependent_spec):
```

**Comment by tgamblin** [comment]
@weijianwen: I think you could update this PR slightly to fix your issue.

**Comment by davydden** [comment]
@Exteris ping.

File: var/spack/repos/builtin/packages/intel-parallel-studio/package.py [hunk 2]
```diff
@@ -347,3 +354,17 @@ def setup_environment(self, spack_env, run_env):
...
+
+    def setup_dependent_package(self, module, dep_spec):
+        if (self.spec.satisfies('+all') or self.spec.satisfies('+mpi')):
+            # Check for presence of bin64 or bin directory
+            if os.path.isdir(self.prefix.bin):
+                bindir = self.prefix.bin
+            elif os.path.isdir(join_path(self.prefix, 'bin64')):
+                bindir = join_path(self.prefix, 'bin64')
+            else:
+                raise "No suitable bindir found"
```
Comment (davydden) [inline var/spack/repos/builtin/packages/intel-parallel-studio/package.py]: in another PR, @adamjstewart said that this should probably be `raise RuntimeError('No suitable bindir found')`

**Comment by davydden** [comment]
superseded by https://github.com/LLNL/spack/pull/3905/

**Pull request closed without merging**